keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,527 @@
+* Separate log message from its arguments
+* Treat log arguments as key-value pairs
+* Be easily parsable and queryable
+* Have a specific guidance on log message and its arguments
+
+For this purpose we suggest to use following log message structure:
+```
+<message> <key1>=<value1> <key2>=<value2> ...
+```
+
+where
+* message is formatted using `%q` fmt logic
+* keys are formatted using `%s` fmt logic
+* values are formatted using `%q` fmt logic, except for number and boolean
+* keys are separated from values with a "=” sign
+* message and key-values are separated with a single space
+
+#### New klog methods
+
+To enforce a new log message structure we want to introduce new methods to klog library which will provide a more structured interface for formatting log messages compared to current methods based on fmt format strings.
+For each format method (`Infof`, `Errorf`) we will add matching structured method (`InfoS`, `ErrorS`). Interface for this methods will model  https://github.com/go-logr/logr (suggested by @thockin)
+Each of those methods will accept log messages as a first argument and a list of key-values pairs as a variadic second argument.
+This approach allows incremental adoption of structured logging without converting ALL of kubernetes to a new API at one time.
+
+Declaration:
+
+```go
+package klog
+
+func InfoS(msg string, keysAndValues ...interface{})
+func ErrorS(err error, msg string, keysAndValues ...interface{})
+```
+
+Example:
+
+```go
+klog.InfoS("Pod status updated", "pod", "kubedns", "status", "ready")
+```
+
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kubedns" status="ready"
+```
+
+And
+
+```go
+klog.ErrorS(err, "Failed to update pod status")
+```
+
+That would result in log
+```
+E1025 00:15:15.525108       1 controller_utils.go:114] "Failed to update pod status" err="timeout"
+```
+
+### References to Kubernetes objects
+
+To address problem of inconsistency in referencing of Kubernetes objects, new klog methods will include additional logic for handling their formatting. Kubernetes objects will be recognized by checking for `ObjectMeta` structure.
+
+Namespaced objects will be represented by joining their name and namespace with `/` character. Pattern: `<namespace>/<name>`, example for pod: `kube-system/kubedns`
+
+Non-namespaced objects will be represented by their name. Pattern: `<name>`, example for node: `cluster1-vm-72x33b8p-34jz`
+
+Example:
+```go
+pod := corev1.Pod{Name: "kubedns", Namespace: "kube-system", ...}
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready")
+```
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+```
+Selecting key under which put objects will be left to caller as defining log schema will need more discussion and is outside of scope of this KEP.
```

Comment:
if you're changing all the lines already, it seems like you'd want to know where you want to go.  Otherwise you'll be doing this sweep again.

Reply:
Managing keys requires solving problems like schema, versioning etc.
Doing it in one KEP would greatly delay any improvements. 

In this KEP we will focus on improving format.

Reply:
This can be included in logging guide, which will be updated as part of this effort.